### PR TITLE
Adding Kaleido

### PR DESCRIPTION
**Pull request by jonmmease**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by jonmmease** [comment]
Hi @conda-forge/help-python-c! This recipe is ready for review, and I wanted to provide a little more context since the structure here may be a little bit unusual.

[Kaleido](https://github.com/plotly/Kaleido) is a library that we (the plotly team) are developing to perform static image export of web-based plotting libraries. This library consists of a standalone C++ command-line executable that embeds [headless chromium](https://chromium.googlesource.com/chromium/src/+/lkgr/headless/README.md). This executable is built following the official Chromium build instructions on Linux, MacOS, and Windows. The Python code launches this executable as a subprocess and communicates with it by writing to standard in, and reading from standard out.

We already have a CI setup to download and compile Chromium and build wheels that are published to PyPI. So my goal here was to build the conda packages from these wheels. Please let me know if you have any questions, and thanks a lot for your time!

**Comment by SylvainCorlay** [comment]
This seems to pull kaleido from a built wheel instead of from source. 
I think that the conda-forge policy is to build everything from source. Would it be possible to build Kaliedo from source?

Also, it seems that kaleido embeds chrome. Could it depdend on a conda package for chrome?

**Comment by jonmmease** [comment]
Hi @SylvainCorlay, thanks a lot for taking a look.

> This seems to pull kaleido from a built wheel instead of from source.
I think that the conda-forge policy is to build everything from source. Would it be possible to build Kaliedo from source?

I'm not sure. On Linux, the Chromium build [instructions](https://chromium.googlesource.com/chromium/src/+/master/docs/linux/build_instructions.md) are tailored to Ubuntu 16.04 and involve setting up an environment with a lot of `apt` installed system dependencies: Most are specified in the [build/install-build-deps.sh](https://github.com/chromium/chromium/blob/master/build/install-build-deps.sh) script.

My understanding is that building from source would require having all of these packaged on conda-forge. Is that correct? I haven't gone through them in detail, but there are a few dozen requirements and, at a glance, there are some like `libnss-dev` and `libxss-dev` that I don't think are on conda-forge right now.

> Also, it seems that kaleido embeds chrome. Could it depdend on a conda package for chrome?

I'm not sure. Would this involve compiling chromium as a shared library and distributing that as a conda package?  The internals of the chromium build process are still a bit opaque to me, but I haven't seen any information on whether it's possible to first compile it as a shared library and then link that against another application.  There are two examples that the Chromium documentation provide of embedding chromium in a custom application (as Kaleido does): `headless_example` and `headless_shell` from https://chromium.googlesource.com/chromium/src/+/lkgr/headless/README.md#usage-as-a-c_library.  Both of these are compiled out of the main Chromium source tree as alternative build targets.

I realize that this approach is not ideal, and I understand that it could introduce problems **_if_** Kaleido were built this way and distributed on conda-forge as a shared library that would get linked into other packages. But since the native portion of Kaleido is compiled as a standalone application, and it is launched as a subprocess of the Python interpreter, my understanding is that there shouldn't be a risk of introducing a binary incompatibility with the rest of conda-forge.

My thinking was that this approach is architecturally similar to the [firefox package](https://github.com/conda-forge/staged-recipes/pull/9234) which downloads the Firefox application tar ball for use as a standalone application. Just that it also includes a pure Python wrapper for launching the application as a subprocess and interacting with it over standard-out / standard-in.

Thanks again!

**Comment by maartenbreddels** [comment]
> But since the native portion of Kaleido is compiled as a standalone application, and it is launched as a subprocess of the Python interpreter, my understanding is that there shouldn't be a risk of introducing a binary incompatibility with the rest of conda-forge.

I agree with this, I think for these use cases there is no reason to rebuild.

**Comment by SylvainCorlay** [comment]
The thing is that all packages of conda-forge are built by conda-forge (presumably), and bringing some binary package breaks that assumption.

I agree that building chrome may be complex, but it would be a huge win for conda-forge to have a well-built chrome package.

**Comment by maartenbreddels** [comment]
But those concerns only matter when linking or loading libraries right, not some 'external' executable, or do you forsee problems that I'm not considering?

**Comment by jonmmease** [comment]
It was pointed out in Gitter (sorry I lost the actual thread), that the Qt feedstock (https://github.com/conda-forge/qt-feedstock) includes support for `QWebEngine` which embeds Chromium.  Presumably this means that it should be possible to build Chromium itself, but the recipe includes over a dozen web-engine related patches (https://github.com/conda-forge/qt-feedstock/tree/master/recipe/patches), so it seems like it might not be straight forward to adapt it.

> but it would be a huge win for conda-forge to have a well-built chrome package.

@SylvainCorlay, could you expand a little on how you see this benefiting the conda-forge ecosystem? I can imagine there eventually being conda packages for full chromium, and for a headless chromium shell.  And then there are projects that embed chromium like Qt, Electron, and Kaleido. But it's not clear to me how Qt/Electron/Kaleido could reuse anything from the full chromium or headless shell recipes (apart from being able to copy/paste parts of the recipe).

It looks like Electron used to use an unsupported process to first build Chromium as a shared library to link to (https://github.com/electron/libchromiumcontent), but it became incompatible as Chromium advanced (See https://www.electronjs.org/blog/electron-internals-building-chromium-as-a-library).

Would it make sense to create a conda package that includes the chromium source for a specific version, and has already been compiled in a particular build configuration (e.g. a headless release build)? It would probably end up being a 20-30GB package. Are there any other conda packages that work kind of like this?

**Comment by jonmmease** [comment]
Wanted to follow up on this @conda-forge/help-python-c.  We're planning to make Kaleido a dependency of the `plotly` package in the not too distant future, and so we'd really like to work out a way for this to work on conda-forge.

For anyone who hasn't followed the whole thread, here's a summary:
 - Kaleido consists of a pure Python layer that launches a platform dependent executable as a subprocess and interacts with it by writing to standard in and reading from standard out.
 - This executable is a custom C++ utility that bundles headless chromium (https://chromium.googlesource.com/chromium/src/+/lkgr/headless/README.md), and it is built following the upstream chromium build instructions.
 - While it isn't an ideal situation to include an executable built outside of conda-forge, Kaleido's architecture prevents the possibility of introducing binary compatibility issues with other conda-forge packages. This is because the native portion of kaleido is launched as a subprocess and not loaded as a shared library.
 - My understanding is that this rationale is also why it's acceptable (if not desired) for the [firefox feedstock](https://github.com/conda-forge/firefox-feedstock) to download and repackage Firefox binaries rather than building them from source.

Based on success of the [qt feedstock](https://github.com/conda-forge/qt-feedstock), it should in principle be possible for Kaleido to be built inside the conda-forge ecosystem. But this would be a substantial effort (both initially and in ongoing maintenance), and at this point I don't see any benefit that would be gained for either Kaleido or conda-forge users. Is there anything I'm missing?

It would be great if it were possible to build chromium as a shared library, and then have packages like Kaleido, Qt, and Electron build against it.  Electron used to have a custom workflow for doing this (https://github.com/electron/libchromiumcontent), but my understanding is that it was deprecated due to the impracticality of keeping up with Chromium development. Some of the difficulties are discussed in https://www.electronjs.org/blog/electron-internals-building-chromium-as-a-library, and this was before the approach was deprecated.

Thanks for considering!

---
According to discussion on Gitter, the lint error here should be fixed by https://github.com/conda-forge/conda-smithy/pull/1383.

**Comment by jonmmease** [comment]
@conda-forge-admin please rerender

**Comment by jonmmease** [comment]
@conda-forge/help-python-c @conda-forge/help-python Friendly ping requesting feedback on https://github.com/conda-forge/staged-recipes/pull/12093#issuecomment-672991369

**Comment by jonmmease** [comment]
Another friendly ping for feedback on https://github.com/conda-forge/staged-recipes/pull/12093#issuecomment-672991369  @conda-forge/help-python-c @conda-forge/help-python. 
Thanks!

**Comment by h-vetinari** [comment]
Building & depending on (variants) of chromium sounds like something that @conda-forge/core should be aware of, escpecially since plot.ly is an important viz package in the ecosystem.

**Comment by beckermr** [comment]
@ericdill has bumped core on this PR. I think we are waiting on someone from @conda-forge/core with the correct expertise to chime in. I don't personally have the correct expertise.

**Comment by scopatz** [comment]
Yeah, we have allowed binary repacks of things like firefox in the past. I think that would/could be acceptable for chromium.  However, it would be best really if there was a `chromium-static` package that this build of kaleido and other projects could build off of. It should be possible to split kaleido from its deps and not redistribute a kaleido binary, and instead build kaleido here.  We'd really want to minimize repackaging of dependencies to just chromium, if at all possible.

**Comment by jonmmease** [comment]
Thanks for the bump @h-vetinari, and for weighing in @scopatz. 

> Yeah, we have allowed binary repacks of things like firefox in the past. I think that would/could be acceptable for chromium.

Thanks, in the short term we'd very much appreciate it!

> However, it would be best really if there was a chromium-static package that this build of kaleido and other projects could build off of. It should be possible to split kaleido from its deps and not redistribute a kaleido binary, and instead build kaleido here. We'd really want to minimize repackaging of dependencies to just chromium, if at all possible.

Yeah, I totally agree in principle here. It's just a bit beyond my packaging ability at this point.  The Chromium project itself doesn't support building as a static library so it would require developing an unofficial process for this. The closest thing I've found is the now-deprecated [libchromiumcontent](https://github.com/electron/libchromiumcontent) that was formerly used by the Electron project. There is some additional info on how this approached used to work at https://www.electronjs.org/blog/electron-internals-building-chromium-as-a-library.

If someone is interested in taking this on and developing a `chromium-static` package down the road, I'd definitely be open to porting the kaleido package to build against it.

**Comment by scopatz** [comment]
Is there a way to build the non-chromium parts of Kaleido here?

**Comment by jonmmease** [comment]
> Is there a way to build the non-chromium parts of Kaleido here?

Not that I know of. The instructions for using Chromium headless as a C++ library are to build the application from within the Chromium source tree. Kaleido basically works just like the `headless_example` app from https://chromium.googlesource.com/chromium/src/+/lkgr/headless/README.md#usage-as-a-c_library. 

Here are C++ source files for kaleido: https://github.com/plotly/Kaleido/tree/master/repos/kaleido/cc.  It's basically one entry point `.cc` file, like `headless_example`, with a handful of `.h` header files that all have includes that reference chromium namespaces.

File: conda-forge.yml [hunk 1]
```diff
@@ -1,2 +1,2 @@
-max_py_ver: '37'
-max_r_ver: '35'
+max_py_ver: '38'
+max_r_ver: '36'
```
Comment (scopatz) [inline conda-forge.yml:2]: Please don't modify files outside of the recipe dir.

Comment (scopatz) [reply conda-forge.yml:2]: Instead use version constraints in the recipe itself.  Though this is a general change that we probably should make

**Comment by scopatz** [comment]
Well, how is this wheel getting created is my question? I am mostly on board with not building chromium here and doing a repack. However, we shouldn't repack a wheel.  We should go from the python source and rebuild the package on CI

**Comment by xhochy** [comment]
Looking at the Kaleido repo, it seems that for Mac (which looks a bit simpler than Linux here), the main script is https://github.com/plotly/Kaleido/blob/c98e92553786b5c450dc1130a8ea27be019af8a1/repos/mac_scripts/build_kaleido ? We should be able to also run that script here in conda-forge. For example, we also have `gn` packaged here on conda-forge and "maybe" could reuse that in the builds. 

I understand that the complexity is smaller when simply doing a binary repack but adopting these build scripts a bit for a conda setup should definitely lead to a much better user experience here on conda-forge. Given the long list in https://github.com/chromium/chromium/blob/master/build/install-build-deps.sh#L309-L346 though, this will be quite some effort. Given that this is an enormous effort to replicate in conda-forge, I would vote be fine with binary repacking for now given that all the licenses are included. Stil, this is just my personal opinion. As others already state, we should discuss this in tomorrow's core meeting.

**Comment by jonmmease** [comment]
> Well, how is this wheel getting created is my question? I am mostly on board with not building chromium here and doing a repack. However, we shouldn't repack a wheel. We should go from the python source and rebuild the package on CI

Right now, the wheel is built in the same CI job where chromium is compiled.  Here's an alternative approach that I could do.

  - Clone the kaleido repo at the tag that corresponds to the recipe version. e.g. for 0.0.3, use https://github.com/plotly/Kaleido/tree/v0.0.3.
  - wget the kaleido binary for the current OS from the GitHub release. e.g. for 0.0.3, download the binary artifact from https://github.com/plotly/Kaleido/releases/tag/v0.0.3.
  - Build the wheel with `python setup.py bdist_wheel ...`

Would this be preferable? If so, I'm happy to update the recipe with this approach. 

Thanks for the help on this!

**Comment by jonmmease** [comment]
Thanks for taking a look @xhochy.  Let me know if there's anything that would be helpful from me for discussion with the core team.

From my understanding, there are two large challenges here.
  1. To build chromium from source in the conda-forge ecosystem would require gn and the ability to reproduce all of the dependencies in https://github.com/chromium/chromium/blob/master/build/install-build-deps.sh#L309-L346 using conda packages. Although, these probably aren't all strictly required when building chromium in headless mode on Linux (which is what kaleido does). It would probably also be desirable to separately package `depot_tools` (https://chromium.googlesource.com/chromium/src/+/master/docs/linux/build_instructions.md#install), which is used to actually download the chromium source tree and handle things like submodules for the current build configuration. Also, expect this to take > 5 hours on CI.

 2. After completing (1), it would be possible to package the full (or headless) Chromium browser in conda-forge. But there's still the question of how the kaleido package would make use of this because there's no upstream-supported way to build chromium as a shared library to link into other applications. As I noted above, Electron use to have a custom downstream patchset for doing this (https://github.com/electron/libchromiumcontent), but they deprecated it a few years ago after facing a lot of challenges in keeping up with upstream.

File: conda-forge.yml (hunk 1 above)
Comment (jonmmease) [reply conda-forge.yml]: Thanks, done in https://github.com/conda-forge/staged-recipes/pull/12093/commits/6d503160575bfaa094611e0fb957fcde31b6d6e7

**Comment by xhochy** [comment]
We had a discussion yesterday at the conda-forge core meeting yesterday and also reached out to you, @jonmmease, separately but I wanted to give some more content here. There are various things that concern us with the repackaging of the wheels.

- The Linux wheels are built for `manylinuxX` specifications and contain shared libraries. The compilers used for `manylinuxX` use a different C++ ABI than the compilers on conda-forge use. Both ABIs should be able to co-exist in the same process but due to peculiarities of the devtoolset compilers in the manylinux Docker images, this can still lead to problems, e.g. the issue where `tensorflow` segfaulted when you imported it after you imported `pyarrow` but not the other way round. I did write a bit about my horrors in https://uwekorn.com/2019/09/15/how-we-build-apache-arrows-manylinux-wheels.html (no need to read it). 
- The wheels vendor many third-party shared libraries as shared objects (`libnssdbm3.so`, `libnssutil3.so`, `libsoftokn3.so`, `libplc4.so`, `libnss3.so`, `libfreeblpriv3.so`, `libfreebl3.so`, `libplds4.so`, `libnssckbi.so`, `libnspr4.so`). These should come from the respective conda packages as they would otherwise conflict with them. This is not so much of an issue for Kaleido but this would probably generate a lot of new bugs in feedstocks that uses these libs and are co-installed with Kaleido by some users.
- Building chromium as a shared library is not that important as maybe previously stated. We would be fine with it being statically linked. It is important though to ship all licenses of bundled third-party libraries. Chromium solves this by having a dialogue where one can view these licenses, this is not an available option for a headless version.
- In addition, building from source will enable you to also ship Kaleido on all platforms conda-forge supports. Currently this is only osx-arm64 and ppc64le in addition to what you can ship on PyPI but we are quite often a bit faster in providing support for popular new platforms as we have the control over the whole toolchain.
- From my personal experience building the wheels and conda packages for `pyarrow`, I can also recommend to use the conda / conda-forge infrastructure to iterate on the packaging / build process. I'm getting much faster feedback and better error message here that in the complicated, hand-crafted docker containers we use for the `pyarrow` wheels.

As with the scripts for building the wheels, this won't be straight-forward but it will be a worthwhile investment to get the plotly stack with this working as flawless as possible in conda-forge, this will hopefully pay off in a reduced number of bug reports.

Happy to work with you here on the PR on improving the build scripts. Feel free to use CI resources to iterate on the scripts, especially as this enables review / commenting on them so we can help you make progress.

**Comment by jonmmease** [comment]
First off, thanks very much to the conda-forge team for taking the time to discuss this, and for reaching out. I'm looking forward to discussing this in a higher bandwidth meeting soon.

@xhochy, thanks for the detailed description of considerations/concerns. I totally understand that these are very important for libraries like pyarrow and qt to avoid binary incompatibility nightmares, but I wanted to emphasize what I believe is an important distinction in kaleido's architecture:

> Both ABIs should be able to co-exist in the same process but due to peculiarities of the devtoolset compilers in the manylinux Docker images ...

Kaleido doesn't load anything into a shared process. The native portion of kaleido (the custom chromium executable and vendored shared libraries) is a standalone command-line executable ("standalone" architecturally, but only really intended to be used programmatically). The Python portion of kaleido is pure Python (nothing compiled, OS independent) that launches the native kaleido executable in a separate process using the [`subprocess`](https://docs.python.org/3/library/subprocess.html) module and communicates with it by writing JSON to standard in, and reading JSON responses from standard out.

So the wheels for the different operating systems all contain identical `*.py` files, they just bundle a different version of the compiled executable.

My understanding is that since the sub process doesn't have a shared memory space with the parent process, there is not risk of introducing a binary incompatibility.  This was a large part of my motivation for adopting this unorthodox architecture, but please let me know if I'm not understanding something properly here.  Thanks!

**Comment by xhochy** [comment]
> Kaleido doesn't load anything into a shared process. The native portion of kaleido (the custom chromium executable and vendored shared libraries) is a standalone command-line executable ("standalone" architecturally, but only really intended to be used programmatically). The Python portion of kaleido is pure Python (nothing compiled, OS independent) that launches the native kaleido executable in a separate process using the [`subprocess`](https://docs.python.org/3/library/subprocess.html) module and communicates with it by writing JSON to standard in, and reading JSON responses from standard out.

Oh yes, this makes it much simpler. Still, the vendored shared libraries are only suboptimal we should not bundle them. With the executable being independent of the Python version, it would actually be preferrable to build them separately on conda-forge. It seems like the Python part is solely pure Python, so the Python package could be a `noarch: python` package that would only be built once and not for the full matrix "platform x Python version". The binary part could be in a `kaleido-bin` package that is just build per platform without the need to expand the matrix by the Python version.

**Comment by jonmmease** [comment]
> With the executable being independent of the Python version, it would actually be preferrable to build them separately on conda-forge. It seems like the Python part is solely pure Python, so the Python package could be a noarch: python package that would only be built once and not for the full matrix "platform x Python version". The binary part could be in a kaleido-bin package that is just build per platform without the need to expand the matrix by the Python version.

This would take a small bit of rework in kaleido itself, but yeah I definitely see the advantages and would be happy to make this change.

> Still, the vendored shared libraries are only suboptimal we should not bundle them.

Ok, yeah, it looks all of the shared library files that we vendor on linux are available across the `expat`, `nspr`, `nss`, and `sqlite` conda-forge packages. So, would we make these packages runtime dependencies of the core kaleido package? Does conda set `RPATH` to point to these libraries? Or does it rely on binary relocation (and the step I'm skipping [here](https://github.com/jonmmease/staged-recipes/blob/6c11e26fafb85c59ade09ae24dde8c4508602d4b/recipes/python-kaleido/meta.yaml#L30-L31) :smiley:)

The only exceptions are the [swiftshader](https://github.com/google/swiftshader) shared libraries. These seem to actually be produced by the chromium build process, and they are a bit odd in the fact that they don't show up in `ldd` and they seem to need to be located in a particular relative path from the executable.  So as long as it's ok to consider these as part of chromium itself, then it looks like nothing else would need to be vendored. 

Also, these swiftshader files are the only shared libraries that are currently bundled on non-linux platforms.

If we go this way, I think we'd still need to produce the vendored wheels for use on PyPI, but maybe we change to pulling these shared libraries from the conda packages.  Is this the kind of thing that [conda-press](https://github.com/conda-incubator/conda-press) was designed to help with?

**Comment by xhochy** [comment]
> Ok, yeah, it looks all of the shared library files that we vendor on linux are available across the `expat`, `nspr`, `nss`, and `sqlite` conda-forge packages. So, would we make these packages runtime dependencies of the core kaleido package? Does conda set `RPATH` to point to these libraries? Or does it rely on binary relocation (and the step I'm skipping [here](https://github.com/jonmmease/staged-recipes/blob/6c11e26fafb85c59ade09ae24dde8c4508602d4b/recipes/python-kaleido/meta.yaml#L30-L31) 😃)
> 
> The only exceptions are the [swiftshader](https://github.com/google/swiftshader) shared libraries. These seem to actually be produced by the chromium build process, and they are a bit odd in the fact that they don't show up in `ldd` and they seem to need to be located in a particular relative path from the executable. So as long as it's ok to consider these as part of chromium itself, then it looks like nothing else would need to be vendored.

Personally, as long as they are in a path that is clearly specific to Kaleido, they can stay in the package. If they don't show in `ldd` it could be that they are loaded at runtime with `dlopen` depending on the graphics drivers available on a system.

> If we go this way, I think we'd still need to produce the vendored wheels for use on PyPI, but maybe we change to pulling these shared libraries from the conda packages. Is this the kind of thing that [conda-press](https://github.com/conda-incubator/conda-press) was designed to help with?

Yes, it should. Not sure whether the Linux builds are compatible with the manylinux specification (might need some tweaks in the visibility flags) but otherwise it should work. @scopatz probably knows better here.

**Comment by xhochy** [comment]
> Ok, yeah, it looks all of the shared library files that we vendor on linux are available across the `expat`, `nspr`, `nss`, and `sqlite` conda-forge packages. So, would we make these packages runtime dependencies of the core kaleido package? Does conda set `RPATH` to point to these libraries? Or does it rely on binary relocation (and the step I'm skipping [here](https://github.com/jonmmease/staged-recipes/blob/6c11e26fafb85c59ade09ae24dde8c4508602d4b/recipes/python-kaleido/meta.yaml#L30-L31) 😃)

`conda` takes care of binary relocation 🥰

**Comment by jonmmease** [comment]
Ok, here's a proposal of things I understand how to do that we can discuss when we meet up.

 - Update the kaleido CI job that compiles kaleido (embedding chromium) to produce a minimal archive that doesn't vendor anything. This would amount to removing the `lib` directory of shared libraries and removing the font files.
 - Create a conda-forge `kaleido`/`kaleido-core`/`kaleido-bin` package that:
   - Declares conda-forge depedencies for these shared libraries and some basic fonts (maybe `fonts-conda-forge`?)
   - Downloads the minimal kaleido executable from the appropriate GitHub release page. 
   - On activation, sets some new special `KALEIDO_EXEC_PATH` environment variable to the path of the kaleido executable.
 - Update the Python portion of the kaleido library to look for `KALEIDO_EXEC_PATH` and, if found, use this path instead of assuming kaleido is vendored in the python package.
 - Update the `python-kaleido` conda-forge package to be `noarch: python` and depend on the `kaleido` package above. This package would checkout the kaleido repo at the appropriate tag, and include only the pure Python portion of kaleido.

I am comfortable with all of these changes.  I see the value of them and, after making these updates, I don't expect there to be much added maintenance burden.  We could either keep our current PYPI pipeline as-is, or update that to pull from the conda pipeline.

I'm less comfortable with the idea of trying to build the core kaleido executable from within the conda forge ecosystem.  It's an intimidatingly complex build system (at least to me) and I'd be very concerned that it would be very fragile and break whenever we wanted to update chromium.  Given that the architecture shouldn't actually cause any compatibility problems with the rest of conda-forge, I'd much rather stick with the [official chromium build instructions](https://chromium.googlesource.com/chromium/src/+/master/docs/linux/build_instructions.md) for maintainability.

That said, if someone really want to try to figure this out at some point, I'd recommend experimenting with the `headless_shell` example described here (https://chromium.googlesource.com/chromium/src/+/lkgr/headless/README.md#usage-as-a-c_library). In terms of architecture and dependencies, Kaleido is just a variation of the `headless_example`/`headless_shell` examples here.

**Comment by scopatz** [comment]
> Is this the kind of thing that conda-press was designed to help with?

Yeah this is the problem that conda-press was meant to solve.  I haven't had any time to keep up with it, though, so I would still consider it in beta.  I will happily review PRs, cut releases, etc.

**Comment by xhochy** [comment]
> * Update the kaleido CI job that compiles kaleido (embedding chromium) to produce a minimal archive that doesn't vendor anything. This would amount to removing the `lib` directory of shared libraries and removing the font files.
> * Create a conda-forge `kaleido`/`kaleido-core`/`kaleido-bin` package that:
>   
>   * Declares conda-forge depedencies for these shared libraries and some basic fonts (maybe `fonts-conda-forge`?)
>   * Downloads the minimal kaleido executable from the appropriate GitHub release page.

I would rather build the kaleido binary inside the conda package. You probably can re-use most of the script you're using in the CI currently. This will though ensure that you build with the same library versions you actually want to link to later. They might be different from the ones you have in your current Ubuntu / CentOS (?) Docker container.

>   * On activation, sets some new special `KALEIDO_EXEC_PATH` environment variable to the path of the kaleido executable.
> * Update the Python portion of the kaleido library to look for `KALEIDO_EXEC_PATH` and, if found, use this path instead of assuming kaleido is vendored in the python package.

Install this binary into `$PREFIX/bin` and look in your Python package (at least in the conda version) for the binary on the `$PATH`. There should be no need for a separate variable and special logic. 

> * Update the `python-kaleido` conda-forge package to be `noarch: python` and depend on the `kaleido` package above. This package would checkout the kaleido repo at the appropriate tag, and include only the pure Python portion of kaleido.

👍 


> I'm less comfortable with the idea of trying to build the core kaleido executable from within the conda forge ecosystem. It's an intimidatingly complex build system (at least to me) and I'd be very concerned that it would be very fragile and break whenever we wanted to update chromium. Given that the architecture shouldn't actually cause any compatibility problems with the rest of conda-forge, I'd much rather stick with the [official chromium build instructions](https://chromium.googlesource.com/chromium/src/+/master/docs/linux/build_instructions.md) for maintainability.
> 
> That said, if someone really want to try to figure this out at some point, I'd recommend experimenting with the `headless_shell` example described here (https://chromium.googlesource.com/chromium/src/+/lkgr/headless/README.md#usage-as-a-c_library). In terms of architecture and dependencies, Kaleido is just a variation of the `headless_example`/`headless_shell` examples here.

It might be intimidating but from my experience (also with `libv8`), this will overall be less work. Happy to help along if you run into issues.

**Comment by jonmmease** [comment]
I did a quick scan through the ubuntu packages installed by the https://chromium.googlesource.com/chromium/src/+/master/build/install-build-deps.sh script.  Here's my best guess on which packages are already on conda forge. I skipped those with identical names that I was pretty certain about:

#### Probably there:
 - libcairo2-dev (cairo is there)
 - libdrm-dev (libdrm-devel-cos7-x86_64 is there)
 - libcap-dev (libcap there)
 - libffi-dev (libffi there)
 - libgbm-dev (mesa-libgbm-cos6-x86_64 there)
 - libglib2.0-dev ( glib2-devel-cos6-x86_64 there)
 - libgtk-3-dev (gtk3 there)
 - libkrb5-dev (krb5 there)
 - libnspr4-dev (nspr v4 there)
 - libpulse-dev (pulse there)
 - libudev-dev (libudev-cos6-x86_64)
 - libxss-dev (libxscrnsaver-cos6-x86_64 there) 
 - libxt-dev (xorg-libxt there)
 - libxtst-dev (xorg-libxtst there)
 - rpm (rpm-tools there)
 - uuid-dev (libuuid-devel-cos7-ppc64le)
 - xz-utils (xz there)
 - libcups2-dev (libcups is there)

#### Maybe
 - dbus-x11 (dbus is there)
 - fakeroot (there is fakeroot-ng, "Fakeroot-ng is a clean re-implementation of fakeroot")
 - git-core (git is there)
 - libbz2-dev (maybe same as bz2file?)
 - libcurl4-gnutls-dev (gnutls and libcurl are there)
 - libelf-dev (no libel either, elfutils is there)
 - libglu1-mesa-dev (libglu there)
 - libpci-dev (libpciaccess there)

#### I don't think so
 - cdbs
 - dpkg-dev (no dpkg either)
 - devscripts
 - libappindicator3-dev (no libappindicator3 either) 
 - libasound2-dev (no libasound2 or asound either)
 - libatspi2.0 (no libatspi or atspi either)
 - libbrlapi-dev (no libbrlapi or brlapi either)
 - libc6-dev (no libc6 or c6 either)
 - libevdev-dev (no libevdev either. evdev under auto channel)
 - libpam0g-dev (no pam0g either)
 - libsctp-dev (no libsctp or sctp either)
 - libspeechd-dev (no libspeechd or speechd either)
 - libwww-perl (bioconda channel has perl-libwww-perl)
 - openbox
 - patch
 - wdiff
 - x11-utils
 - xcompmgr

**Comment by ericdill** [comment]
Just got off a zoom call with @jonmmease @wolfv and @mbargull . Notes are [here](https://hackmd.io/ZDLgWo6pSEaE57RYiWj3Vg?edit) The plan that we've come up with is:

* repackage the wheel
* adding the centos 7 compat to the build section, add `- sysroot_linux-64 >=2.17  # [linux]` in the `requirements/build`.
* make sure shared libraries are getting pulled from CF, specifically: expat, nss, nspr, sqlite. 
    * Fallback option is to continue to vendor these libs, but will need to make sure to package the licenses there too. This won't collide with other packages. The kaleido binary sets the RPATH to the local libraries. They're currently vendored inside of 
    * e.g., put into ${ENV_PATH}/share/kaleido/* instead of ${ENV_PATH}/lib/pythonX.X/site-packages/kaleido/lib/*
    * (J) Can we set an env var for this path using the activate/deactivate scripts in conda?
    * (M) Use an RPATH instead. If you put the kaleido binary (or the wrapper script) into the ${ENV_PATH}/bin (or ${ENV_PATH}/scripts on win), then you already have the binary on path so you dont have to do anything special
* macos
    * when you compile with the CF stack, you get a minimum target (build for macos 10.9). it's again like the glibc thing on linux. We have a minimum macos version that we support, but you can override it. If it doesn't work with 10.9 then that's ok too and pull it up to whatever version it's compatible with. TODO: conda_build_config.yaml in the recipe you put a key MACOS_DEPLOYMENT_TARGET (or something)
    * When you have to do binary relocation, then there is one compiler flag that lets us write the longer paths (`-Wl,-headerpad_max_install_names`, see ). Need to change the `binary_relocation` in the recipe if the shared libraries aren't vendored any more. Add a test where you run the binary and it should fail if it doesn't find the libraries. 
* win
    * TODO: Is there some sort of minimum version thing here for windows? 
* Other platforms:
    * win32. conda-forge doesn't support win32 any more
    * linux-arm64. open issue on feedstock to build this. chromium build currently cross-compiles these.
* make sure to include licenes for chromium headless
* Other TODOs
    * Open issue on feedstock to move to building on CF CI instead of repackaging wheel

Thanks all!

**Comment by nehaljwani** [comment]
Hello friends!

Recently, I spent a few hours banging my head to get kaleido to work on a machine with glibc 2.12. In case any one is interested, I have posted the build script at https://github.com/plotly/Kaleido/issues/51#issuecomment-702358671

**Comment by beckermr** [comment]
Not stele yet!

**Comment by jonmmease** [comment]
This was superseded by these now merged PRs:
 - `kaleido-core`: https://github.com/conda-forge/staged-recipes/pull/12747
 - `python-kaleido`: https://github.com/conda-forge/staged-recipes/pull/13399

Big thanks again to everyone who helped get this across the finish line!

**Pull request closed without merging**